core/unix/signal.c

```diff
@@ -2921,6 +2920,13 @@ sigcontext_to_mcontext(priv_mcontext_t *mc, sig_full_cxt_t *sc_full,
  * thread_get_mcontext() sets the dcontext isa_mode from cpsr, and thread_set_mcontext()
  * sets it.
  */
+/* XXX: on RISC-V, the tp register is a general-purpose register, but as its name
+ * suggests, it is conventionally used as the thread register.
+ * We save DR's tls by keeping tp's value unchanged, regardless in DR or in application.
```

Comment:
Is tp the stolen register?  But arm is writing the stolen register in these functions, and I think handling it in separate code (like set_sigcxt_stolen_reg()).  Is all code using these mcontext<->sigcontext conversion routines updating tp afterward and that's why it doesn't matter whether it's touched here?

Reply:
It's not the stolen register. Per calling convention, `tp` is the thread-pointer register used to point to the thread-local storage. We also have a stolen register (default to x28) same as ARM/AArch64. The strategy is to keep tp unchanged, i.e. mangle every instruction that uses tp to something like:

```assembly
sd scratch_reg, scratch_slot(x28) ; x28 is the stolen reg
ld scratch_reg, app_tls_slot(x28)

; mangled instruction here (tp to scratch_reg)

sd scratch_reg, app_tls_slot(x28)
ld scratch_reg, scratch_slot(x28)
```

So the only difference is on RISC-V, tp is a general register while on ARM/AArch64, it's a special register, which is naturally unchanged.

And here in the mcontext<->sigcontext conversion, if we do not exclude tp, DR(host)'s tp register will be corrupted.

But I'm not sure if this is the correct way to do it after reading your comment. Could you give some advice?

Reply:
So essentially `tp` is a 2nd stolen register.  For any stolen register we need to translate it to the app value when the kernel presents the `tp` register to the app, such as when passed to a signal handler.  Is that done today?  Everywhere we do things like `set_sigcxt_stolen_reg()` we need to also ensure `tp` has the right value.

I think these mcontext<->sigcontext conversion routines assume the caller will deal with translating stolen registers: they just blindly copy fields from one to the other.  So I would expect them to include `tp`.  The caller has to translate it to/from app/DR values if necessary.

Reply:
On second thought, the difference between tp and stolen_reg is that we cannot restore tp's value as there is no backup for it. And since tp is always pointing to host tls, we can't translate it to the app value in the signal handler, or we will lose its original value (host tls) -- it will be mangled if it's used by any instruction in the app.

Reply:
> we cannot restore tp's value as there is no backup for it. And since tp is always pointing to host tls, we can't translate it to the app value in the signal handler, or we will lose its original value (host tls) 

For transparency we need to translate it to the app value in the signal handler.  We have to add storage of DR's value somewhere in the OS fields (as is done on other arches) (probably it is already stored somewhere today I would think).

Reply:
> (probably it is already stored somewhere today I would think)

Yes, we have `dr_tls_base` in the OS field, but that's *DR tls*, not host tls (I just found out we call host tls `lib_tls_base`, sorry for the confusion). IIUC other arches like AArch64 don't have the problem here, since its `lib_tls_base` is stored in a special register `TPIDR_EL0` and is not affected by signal handler sifgreturn.

On the RISC-V side, how do we resume `lib_tls_base` back to tp if the kernel replaced it with app tls? (since we need to translate it to the app value in the signal handler).

Reply:
Looking again at the callers of `set_sigcxt_stolen_reg()`, do we need to translate `tp` to the `app_tls` at the signal handler and then restore it back to `lib_tls_base` before `sigreturn`? If so, is this the transparency for clients or something?

Reply:
Transparency for the app, as it can examine the machine state passed to its signal handler.  See https://dynamorio.org/transparency.html#sec_trans_pretend and the transparency slides in the DR tutorial.

Reply:
Thanks, I reverted this change and added `set_sigcxt_tp_reg()` in [a4610b7](https://github.com/DynamoRIO/dynamorio/pull/6461/commits/a4610b74b9884c5dcd80cba36a00c18fd72a094a). Also enabled building of `linux.eintr` test, since it's working on real hardware. I think it's better to have another round of review.